File: Line.cpp
#include <sstream>
#include <string>
#include <cmath>
#include <iostream>
#include "Point.h"
#include "Line.h"

/// <summary>
/// Конструктор с параметрами
/// </summary>
/// <param name="begin">2D точка начала отрезка</param>
/// <param name="end">2D точка конца отрезка</param>
Line::Line(const Point& begin, const Point& end) : m_begin{ begin }, m_end(end)
{   // здесь у Point вызывается конструктор копирования
    std::cout << ">Line::Line(&" << begin.toString() << ", &" << end.toString() << ")" << std::endl;
}

//Line::Line(const Point& begin, const Point& end)
//{
//    // если конструктор написать так, то для каждого скрытого поля
//    // сначала вызовется КОНСТРУКТОР ПО УМОЛЧАНИЮ
//    // и только затем произойдёт присваивание
//    std::cout << ">Line::Line(&" << begin.toString() << ", &" << end.toString() << ")" << std::endl;
//    setBegin(begin);
//    setEnd(end);
//}

/// <summary>
/// Деструктор
/// </summary>
Line::~Line()
{
    std::cout << ">Line::~Line()" << std::endl;
}

//void Line::setBegin(const Point& begin)
//{
//    std::cout << ">Line::setBegin(&" << begin.toString() << ")" << std::endl;
//    // здесь у Point вызывается конструктор копирования
//    m_begin = begin;
//}
//
//void Line::setEnd(const Point& end)
//{
//    std::cout << ">Line::setEnd(&" << end.toString() << ")" << std::endl;
//    // здесь у Point вызывается конструктор копирования
//    m_end = end;
//}

/// <summary>
/// Геттер для поля begin
/// </summary>
/// <returns>Константная ссылка на 2D точку начала отрезка</returns>
const Point& Line::getBegin() const
{
    std::cout << ">Line::getBegin()" << std::endl;
    return m_begin;
}

/// <summary>
/// Геттер для поля end
/// </summary>
/// <returns>Константная ссылка на 2D точку конца отрезка</returns>
const Point& Line::getEnd() const
{
    std::cout << ">Line::getEnd()" << std::endl;
    return m_end;
}

/// <summary>
/// Вычислить длину отрезка
/// </summary>
/// <returns>Длина отрезка по координатам двух 2D точек</returns>
double Line::getLenght() const
{
    std::cout << ">Line::getLenght()" << std::endl;
    const Point& begin = getBegin();
    const Point& end = getEnd();
    double result = std::sqrt(
        std::pow(end.getX() - begin.getX(), 2) +
        std::pow(end.getY() - begin.getY(), 2)
    );
    return result;
}

/// <summary>
/// Вывод в стандартный поток вывода
/// </summary>
void Line::display() const
{
    std::cout << ">Line::display()" << std::endl;
    std::cout << toString() << std::endl;
}

/// <summary>
/// Получить текстовое представление объекта
/// </summary>
/// <returns>Текстовое представление объекта</returns>
std::string Line::toString() const
{
    std::cout << ">Line::toString()" << std::endl;
    std::ostringstream os;
    os << "Point(" << getBegin().toString() << ", " << getEnd().toString() << ")";
    return os.str();
}

File: 00-java/src/Line.java
/**
 * 2D отрезок
 */
public class Line {

    /**
     * Конструктор с параметрами
     *
     * @param begin 2D точка начала отрезка
     * @param end   2D точка конца отрезка
     */
    public Line(Point begin, Point end) {
        System.err.println(">Line(" + begin + ", " + end + ")");
        setBegin(begin);
        setEnd(end);
    }

    /**
     * Вычислить длину отрезка
     *
     * @return Длина отрезка по координатам двух 2D точек
     */
    public double getLength() {
        System.err.println(">Line.getLength()");
        Point begin = getBegin();
        Point end = getEnd();
        double x2 = Math.pow(end.getX() - begin.getX(), 2);
        double y2 = Math.pow(end.getY() - begin.getY(), 2);
        return Math.sqrt(x2 + y2);
    }

    /**
     * Получить текстовое представление объекта
     *
     * @return Текстовое представление объекта
     */
    @Override
    public String toString() {
        return "Line(" + getBegin() + ", " + getEnd() + ")";
    }

    /**
     * Инициализатор объекта из стандартного потока ввода
     *
     * @return Сконструированный объект из данных
     */
    public static Line read() {
        System.err.println(">Line.read()");
        System.out.println("Введите координаты начала отрезка:");
        Point begin = Point.read();
        System.out.println("Введите координаты конца отрезка:");
        Point end = Point.read();
        return new Line(begin, end);
    }

    public Point getBegin() {
        System.err.println(">Line.getBegin()");
        return m_begin;
    }

    public Point getEnd() {
        System.err.println(">Line.getEnd()");
        return m_end;
    }

    public void setBegin(Point begin) {
        System.err.println(">Line.setBegin(" + begin + ")");
        m_begin = begin;
    }

    public void setEnd(Point end) {
        System.err.println(">Line.setEnd(" + end + ")");
        m_end = end;
    }

    /**
     * Координаты начала отрезка
     */
    private Point m_begin;

    /**
     * Координаты конца отрезка
     */
    private Point m_end;
}
